File: clojure/leap/src/leap.clj
(ns leap)

(defn evenly-divisble-by
  [dividend divisor]
  (zero? (mod dividend divisor)))

(defn leap-year? [year]
  (if (evenly-divisble-by year 100)
    (evenly-divisble-by year 400)
    (evenly-divisble-by year 4)))

File: kotlin/leap/src/main/kotlin/Leap.kt
data class Year(val year: Int) {
    val isLeap: Boolean by lazy {
        when {
            isEvenlyDivisible(400) -> true
            isEvenlyDivisible(100) -> false
            isEvenlyDivisible(4) -> true
            else -> false
        }
    }

    private fun isEvenlyDivisible(divisor: Int): Boolean = year.mod(divisor) == 0
}
